lib/rubocop/cop/sorbet/binding_constant_without_type_alias.rb

```diff
@@ -0,0 +1,102 @@
+# frozen_string_literal: true
+
+require "rubocop"
+
+module RuboCop
+  module Cop
+    module Sorbet
+      # This cop disallows binding the return value of `T.any`, `T.all`, `T.enum`
+      # to a constant directly. To bind the value, one must use `T.type_alias`.
+      #
+      # @example
+      #
+      #   # bad
+      #   FooOrBar = T.any(Foo, Bar)
+      #
+      #   # good
+      #   FooOrBar = T.type_alias { T.any(Foo, Bar) }
+      class BindingConstantWithoutTypeAlias < RuboCop::Cop::Base
+        extend AutoCorrector
+
+        MSG = "It looks like you're trying to bind a type to a constant. " \
+          "To do this, you must alias the type using `T.type_alias`."
+        WITHOUT_BLOCK_MSG = "It looks like you're using the old `T.type_alias` syntax. " \
+          "`T.type_alias` now expects a block." \
+          'Run Sorbet with the options "--autocorrect --error-white-list=5043" ' \
+          "to automatically upgrade to the new syntax."
+
+        def_node_matcher :type_alias_without_block, <<~PATTERN
+          (send
+            (const {nil? cbase} :T)
+            :type_alias
+            $_
+          )
+        PATTERN
+
+        def_node_matcher :type_alias_with_block?, <<~PATTERN
+          (block
+            (send
+              (const {nil? cbase} :T)
+            :type_alias)
+            ...
+          )
+        PATTERN
+
+        def_node_matcher :requires_type_alias?, <<~PATTERN
+          (send
+            (const {nil? cbase} :T)
+            {
+              :all
+              :any
+              :class_of
+              :nilable
+              :noreturn
+              :proc
+              :self_type
+              :untyped
+            }
+            ...
+          )
+        PATTERN
+
+        def on_casgn(node)
+          expression = node.expression
+          return if expression.nil? # multiple assignment
+
+          type_alias_without_block(expression) do |type|
+            return add_offense(expression, message: WITHOUT_BLOCK_MSG) do |corrector|
+              corrector.replace(expression, "T.type_alias { #{type.source} }")
+            end
+          end
+
+          return if type_alias_with_block?(expression)
+
+          requires_type_alias?(send_leaf(expression)) do
+            return add_offense(expression, message: MSG) do |corrector|
+              corrector.replace(expression, "T.type_alias { #{expression.source} }")
+            end
+          end
+        end
+
+        private
+
+        # Given nested send nodes, returns the leaf with explicit receiver.
+        #
+        # i.e. in Ruby
+        #
+        #     a.b.c.d.e.f
+        #     ^^^
+        #
+        # i.e. in AST
+        #
+        #     (send (send (send (send (send (send nil :a) :b) :c) :d) :e) :f)
+        #                             ^^^^^^^^^^^^^^^^^^^^^^^
+        #
+        def send_leaf(node)
+          node = node.receiver while node&.receiver&.send_type?
+          node
+        end
```

Comment:
This was hard to name. I'm still not satisfied with it, but it's the best I've got right now.

I feel like the concept is simple, but the fact that the Ruby code and AST are "inverted" makes the root/leaf/parent terminology awkward.

Long story short

```ruby
T.proc.void
^^^^^^ extracts this
```

I made it a loop, so that it will continue to work even if there's some weird usage like

```ruby
T.proc.foo.bar.baz
```

Reply:
How come we didn't need this in the old implementation?

Reply:
The old implementation was using [this node search](https://github.com/Shopify/rubocop-sorbet/blob/0333a8e25e06890dd100d8313f06ad4995a57cac/lib/rubocop/cop/sorbet/binding_constants_without_type_alias.rb#L79-L87), which meant it would traverse all descendant looking for a match anywhere.

This means it would get confused by the following

```ruby
A = lambda do
  # ^^^^^^^^^ It looks like you're trying to bind a type to a constant. To do this, you must alias the type using `T.type_alias`.
  x = T.let(1, T.untyped) # Because it erroneously detects this `T.untyped`
  x
end
```

while that's unlikely, the new implementation tries to be more explicit about what it's looking for: we only care about the starting `T.something` method send.